File: OSC/OSCserver.pl
#!/usr/bin/perl

use strict;
use Net::OpenSoundControl::Server;
my $port = 4567;

use Data::Dumper qw(Dumper);

sub ledAction {
   	#re-initialize each variable for each cal to ledAction 
    my $color   = '';
    my $rpiCmd  = '';
    my $pltnCmd = '';
    my $srvcCmd = '';
    
    my $oscMsgRef  = $_[1]; #assign OSC msg reference to var
    my $oscAddress = "@{ $oscMsgRef }[0]";
    my $oscType    = "@{ $oscMsgRef }[1]";
    my $oscValue   = "@{ $oscMsgRef }[2]";
    #print Dumper $oscMsgRef;
    print "Address:" . "$oscAddress\n" ;
    print "Type:" . "$oscType\n" ;
    print "Value:" . "$oscValue\n\n";
    
    #parse commands from OSC addresses
    $oscAddress =~ m/\/led\/(\w+)/;
    $color = $1;

	$oscAddress =~ m/\/rpi\/(\w+)/;
	$rpiCmd = $1;
    
	$oscAddress =~ m/\/pltn\/(\w+)/;
	$pltnCmd = $1;
	
	$oscAddress =~ m/\/srvc\/(\S+)/;
	$srvcCmd = $1;
	
    #hash of colors and corresponding GPIO on RPi
    my %ledGPIO = (
	     red1   => "2", #GPIO 18 
		 green1 => "5", #     23
		 blue1  => "7", #     25
		 red2   => "1", #     17
		 green2 => "4", #     22
		 blue2  => "6", #     24
	);
    # LED (/led) OSC commands 
    if ($oscAddress =~ 'led') {
    	#Set to all white
    	if ($color eq 'white') {
    		my $colorPerc = 0.00001;
			foreach my $key ( keys %ledGPIO )
			{ 
    			system("echo \"$ledGPIO{$key}=$colorPerc\" > /dev/pi-blaster"); 
   			}	
    	#Turn all LEDs off
    	} elsif ($color eq 'allOff') {
			foreach my $key ( keys %ledGPIO )
			{ 
    			system("echo \"$ledGPIO{$key}=0\" > /dev/pi-blaster"); 
   			}	
    	# set LED color according to passed pin value
    	} else {
    		system("echo \"$ledGPIO{$color}=$oscValue\" > /dev/pi-blaster"); 
    	}
    }
    
    #Handle services (/srvc) OSC commands
    if ($oscAddress =~ 'srvc') {
    	my $sysCmd = '';
    	if ($oscValue == 1){
    		$sysCmd = "sudo service " . $srvcCmd . " start"; 
    	} elsif ($oscValue == 0){
    		$sysCmd = "sudo service " . $srvcCmd . " stop"; 
    	} else{
		   	print "not a valid /srvc OSC address";	
    	}
    	system $sysCmd;	
    	print "sysCmd: " . $sysCmd ."\n";
    }
    
	#Handle RPi (/rpi) PSC Commands
	if ($oscAddress =~ 'rpi') {
    	my $sysCmd = '';
   		if ($rpiCmd eq 'off') {
			print "Shutting Down the Raspberry Pi";
			for (my $i = 0; $i <= 5; $i++) {
				print ".";
				sleep (1);
			}
			$sysCmd = 'sudo shutdown -h now';
    	} else {
    		print "Not a valid RPi OSC command\n"
    	}
    	#run the system command
   		system $sysCmd;
	}  
}

my $server =
    Net::OpenSoundControl::Server->new(Port => $port, Handler => \&ledAction)
    or die "Could not start server: $@\n";

print "[OSC Server] Receiving messages on port $port\n";

$server->readloop();
print $server->port();

File: OSC/OSCServer.py
#!/usr/bin/python
"""@package OSCServer
Documentation for this module
"""

import sys
sys.path.append("/usr/local/pltn/Modules")
from OSC import OSCServer
from LED import CommonLED
from LED import effects
import time
from subprocess import call
import threading
from pltnGpio import pltnGpio

#import ledTwitter

#dict for GPIO pins
pg = pltnGpio.pltnGpio()
gpioPins = pg.getAllPins('asDict')
ef = effects.effects()

#Define OSC server port and traceback IP
OSCPort = 4567
OSCIP   = "0.0.0.0"

#Instantiate server
oscSrv = OSCServer((OSCIP,OSCPort))
#instantiate time object
localtime = time.asctime( time.localtime(time.time()) )
#instantiate common LED object
cLED = CommonLED.CommonLED()

def led(path, tags, args, source):
	"""Callback function to handle all LED functions.
		OSC Msg: /pltn/led <color+stripNum>|<LEDprogram> <action> (0|1)
		Examples: /pltn/led r1 1 solid
		          /pltn/led b1 1 flash
		          /pltn/led g1 1 flashFade
	"""
	oscProg  = args[0]
	#pinValue = args[1]	
	#action   = args[2]
	print oscProg
	#print pinValue
	#print action
	
	#check if first argument is a pin value
	if oscProg in gpioPins.keys():
		pinValue = args[1]
		action   = args[2]
		#search gpioPins dict for pin value. Exit when found
		for dictColor,gpioPin in gpioPins.iteritems():
			if oscProg == dictColor:
				break
		#set the pin color
		if action == 'solid':
			#cLED.setPinValue(gpioPin,pinValue,0.01)
			cLED.setPinValue(gpioPin,pinValue)
		elif action == 'flashFade':
			#cLED.ledFlashFade(gpioPin,1)
			t = threading.Thread(target=ef.ledFlashFade,args=(gpioPin,pinValue,0.01))
			t.start()
			t.join
		elif action =="flash":
			ef.flash(gpioPin,0.1)
		else:
			#not a valid option
			pass			
		
	#Turn all LEDs on
	elif oscProg == 'allOn':
		cLED.setColor(1,[1,1,1])
		cLED.setColor(2,[1,1,1])
	#Turn all LEDs off
	elif oscProg == 'allOff':
		cLED.allOff()
	else:
		pass	
		
def rpi(path, tags, args, source):
	"""
	Callback function to handle all RPi related functions
		OSC Msg: /pltn/rpi <function> <secretKey>
	"""
	#get the RPi command to run
	authzKey = "3681"
	cmd = args[0]
	key = args[1]
	#Check for proper command and authorization Key
	if cmd == 'off' and key == authzKey:
		print "{0}: RPi received shutdown command" .format(localtime)
		call(["sudo", "shutdown", "-h", "now"])
	else:
		print "{0}: \"{1} {2} {3}\" Not an allowed function/key combo" .format(localtime,path,cmd,key)

def srvc(path, tags, args, source):
	"""
	Callback function to handle all RPi related functions
		OSC Msg: /pltn/srvc <srvcName> start|stop
	"""
	#list of allowed services and values. Security to prevent rogue 
	#msgs being sent and started
	allowedSrvcs = ["pi-blaster","ssh","rangeSensor","pltnAgent"]
	allowedCmds  = ["start","stop","status"]
	srvcName = args[0]
	value    = args[1]
	#check if this is an allowed command
	if srvcName in allowedSrvcs and value in allowedCmds:
		call(["sudo", "service", srvcName, value])
	else:
		print "{0}: \"{1} {2} {3}\" Not allowed" .format(localtime,path,srvcName,value) 

def heartbeat(path, tags, args, source):
	"""
	Callback function to process heartbeats from RPi. 
	Code here will send to OSC server on CC.pde 
	"""
	print "---------------"
	print path
	print args[0]
	print args[1]
	print args[2]
	print args[3]
	
#def tweet(path, tags, args, source):
	#ledTwitter.fadeTweet()

#Message Handlers and Callback functions
oscSrv.addMsgHandler("/pltn/led",led)
oscSrv.addMsgHandler("/pltn/srvc",srvc)
oscSrv.addMsgHandler("/pltn/rpi",rpi)
oscSrv.addMsgHandler("/pltn/heartbeat",heartbeat)
#oscSrv.addMsgHandler("/led/tweet",tweet)

print "\n listening on port: %i" % OSCPort

try:
	while True:
		oscSrv.handle_request()

except KeyboardInterrupt:
	cLED.allOff()
	print "Quit"
	oscSrv.close()
